### ISSUE TITLE
[8.0] records.write() on calendar events only updates the last record

### ISSUE THREAD

**Issue by michaelkarrer81** [odoo/odoo#9544]
## EDIT: 
The initial **ValueError Expected singleton** was caused by me since i had an ensure.one in my inherited .write methods but after i corrected my fault the .write method did not throw an error but would only update one of the records in my recordset! Therefore i reopened the issue but changed the title of the issue.

## Original Post:
I try to update (write) a many2one field to all records of a recordset in one DB write (just like the example in the documentation) but every time i try it this way get an **ValueError Expected singleton**?!?

#### I tried:
```python
events.write({'category_id': self.category_new_id})
```
```python
events.write({'category_id': self.category_new_id.id})
```
but in every case i get an **ValueError Expected singleton** error.

Unfortunately it will work with a simple for loop:
### Working function (but with a for loop instead of a single write):
```python
from openerp import models, api
from openerp.fields import Many2one

class ReplaceCategoryWizard(models.TransientModel):
    _name = 'calendar_log.replace_eventcategory_wizard'

    category_old_id = Many2one('calendar.event.category', string='Old Category', required=True)
    category_new_id = Many2one('calendar.event.category', string='Replaced by Category', required=True)

    @api.multi
    def replace_category(self):
        assert self.category_old_id.id != self.category_new_id.id
        events = self.env['calendar.event'].search([('category_id', '=', self.category_old_id.id)])
        
        #events.write({'category_id': self.category_new_id.id}) # FAILS with ValueError Expected singleton?!
        #events.write({'category_id': self.category_new_id}) # FAILS with ValueError Expected singleton?!
        
        # But this works:
        for event in events:
            event.category_id = self.category_new_id
        return {}
```

### What would be the correct way to update/set the value of a many2one field for all records of a recordset?

**Comment by andreparames** [comment]
Can you post the traceback you get?

**Comment by michaelkarrer81** [comment]
# @andreparames Traceback for way a (which i do understand)

```python
events.write({'category_id': self.category_new_id})
```

```
2015-11-13 11:18:33,128 12809 ERROR o8_exthof_hofermedical openerp.sql_db: Programming error: can't adapt type 'calendar.event.category', in query UPDATE "calendar_event" SET "category_id"=%s,"write_uid"=%s,"write_date"=(now() at time zone 'UTC') WHERE id IN %s
2015-11-13 11:18:33,128 12809 ERROR o8_exthof_hofermedical openerp.http: Exception during JSON request handling.
Traceback (most recent call last):
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 574, in dispatch
    result = self._call_function(**self.params)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/service/model.py", line 118, in wrapper
    return f(dbname, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 307, in checked_call
    return self.endpoint(*a, **kw)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 803, in __call__
    return self.method(*args, **kw)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 403, in response_wrap
    response = f(*args, **kw)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/addons/web/controllers/main.py", line 948, in call_button
    action = self._call_kw(model, method, args, {})
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 372, in old_api
    result = method(recs, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/addons-loaded/calendar_log/wizard/wizard.py", line 20, in replace_category
    events.write({'category_id': self.category_new_id})
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 248, in wrapper
    return new_api(self, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/addons-loaded/calendar_log_project/models/calendar.py", line 67, in write
    res = super(calendar_event, self).write(values)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 248, in wrapper
    return new_api(self, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 565, in new_api
    result = method(self._model, cr, uid, self.ids, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/addons/calendar/calendar.py", line 1613, in write
    res = super(calendar_event, self).write(cr, uid, [int(event_id) for event_id in ids], values, context=context)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/addons/mail/mail_thread.py", line 432, in write
    result = super(mail_thread, self).write(cr, uid, ids, values, context=context)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 372, in old_api
    result = method(recs, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/models.py", line 3787, in write
    self._write(old_vals)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 248, in wrapper
    return new_api(self, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 565, in new_api
    result = method(self._model, cr, uid, self.ids, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/models.py", line 3898, in _write
    cr.execute(query, params + (sub_ids,))
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/sql_db.py", line 158, in wrapper
    return f(self, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
ProgrammingError: can't adapt type 'calendar.event.category'
```

**Comment by michaelkarrer81** [comment]
# @andreparames Traceback for way B which i don't understand
```python
events.write({'category_id': self.category_new_id.id})
```

```
2015-11-13 11:22:07,509 12852 ERROR o8_exthof_hofermedical openerp.http: Exception during JSON request handling.
Traceback (most recent call last):
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 574, in dispatch
    result = self._call_function(**self.params)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/service/model.py", line 118, in wrapper
    return f(dbname, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 307, in checked_call
    return self.endpoint(*a, **kw)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 803, in __call__
    return self.method(*args, **kw)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/http.py", line 403, in response_wrap
    response = f(*args, **kw)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/addons/web/controllers/main.py", line 948, in call_button
    action = self._call_kw(model, method, args, {})
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 372, in old_api
    result = method(recs, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/addons-loaded/calendar_log/wizard/wizard.py", line 20, in replace_category
    events.write({'category_id': self.category_new_id.id})
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/api.py", line 248, in wrapper
    return new_api(self, *args, **kwargs)
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/addons-loaded/calendar_log_project/models/calendar.py", line 69, in write
    if res and self.ensure_one():
  File "/Users/mkarrer/Entwicklung/github/odoo_v8.0/odoo/openerp/models.py", line 5303, in ensure_one
    raise except_orm("ValueError", "Expected singleton: %s" % self)
except_orm: ('ValueError', 'Expected singleton: calendar.event(652, 650, 649, 648, 647, 646, 645, 644, 643, 642, 641, 640, 639, 637, 636, 635, 634, 633, 632, 631, 630, 629, 628, 626, 623, 622, 621, 620, 619, 618, 617, 616, 615, 614, 613, 612, 611, 610, 609, 608, 607, 606, 605, 604, 603, 602, 601, 600, 599, 598, 597, 596, 595, 594, 593, 592, 591, 590, 589, 588, 587, 586, 585, 584, 583, 582, 581, 580, 579, 578, 577, 576, 575, 574, 573, 572, 571, 570, 569, 568, 567, 566, 565, 564, 563, 562, 561, 560, 559, 558, 557, 556, 555, 554, 553, 552, 551, 550, 549, 548, 547, 546, 545, 544, 543, 542, 541, 540, 539, 538, 537, 536, 535, 534, 533, 532, 531, 530, 529, 528, 527, 526, 525, 524, 523, 522, 521, 520, 519, 518, 517, 516, 515, 514, 513, 512, 511, 510, 509, 508, 507, 506, 505, 504, 503, 502, 501, 500, 499, 498, 497, 496, 495, 494, 493, 492, 491, 490, 489, 488, 487, 486, 485, 484, 483, 482, 481, 480, 478, 477, 476, 475, 474, 473, 472, 471, 470, 469, 468, 467, 466, 465, 464, 463, 462, 461, 460, 459, 458, 457, 456, 455, 454, 453, 452, 451, 449, 448, 447, 446, 445, 444, 443, 442, 441, 440, 439, 438, 437, 436, 435, 434, 433, 432, 431, 430, 429, 428, 427, 426, 425, 424, 423, 422, 421, 420, 419, 418, 417, 416, 415, 414, 413, 412, 411, 410, 409, 408, 407, 406, 404, 403, 402, 401, 400, 399, 398, 397, 396, 395, 394, 393, 392, 390, 389, 388, 387, 385, 384, 383, 381, 380, 379, 378, 377, 376, 375, 374, 373, 372, 371, 370, 369, 368, 367, 366, 365, 364, 363, 362, 361, 360, 359, 358, 357, 356, 355, 354, 353, 352, 351, 350, 349, 348, 347, 346, 345, 344, 343, 342, 341, 340, 339, 338, 337, 336, 335, 334, 333, 332, 331, 330, 329, 328, 327, 326, 325, 324, 323, 322, 321, 320, 319, 318, 317, 316, 315, 314, 313, 312, 311, 310, 309, 308, 307, 306, 305, 304, 303, 302, 301, 300, 299, 298, 297, 296, 295, 294, 293, 292, 291, 290, 289, 288, 287, 286, 285, 284, 283, 282, 281, 280, 279, 278, 277, 276, 275, 274, 273, 272, 271, 270, 269, 268, 267, 266, 265, 263, 262, 261, 260, 259, 258, 257, 256, 255, 254, 253, 252, 251, 250, 249, 248, 247, 246, 245, 244, 243, 242, 241, 240, 239, 238, 237, 236, 235, 234, 233, 232, 231, 230, 229, 228, 227, 226, 225, 224, 223, 222, 221, 220, 219, 218, 216, 215, 214, 213, 212, 211, 210, 209, 208, 207, 206, 205, 204, 203, 202, 201, 200, 199, 198, 197, 196, 195, 194, 193, 192, 191, 190, 189, 188, 187, 186, 185, 184, 183, 182, 181, 180, 179, 178, 176, 175, 174, 173, 172, 171, 170, 169, 168, 167, 166, 165, 164, 163, 162, 161, 160, 159, 158, 157, 156, 155, 154, 153, 152, 151, 149, 148, 147, 146, 145, 144, 142, 141, 140, 139, 138, 137, 136, 135, 134, 132, 131, 130, 129, 128, 127, 126, 125, 124, 123, 122, 114, 113, 112, 111, 110, 109, 108, 107, 106, 105, 104, 103, 102, 101, 100, 99, 98, 97, 96, 95, 94, 93, 92, 91, 90, 89, 88, 87, 86, 84, 79, 77, 75, 73, 71, 69, 66, 64, 62, 60, 59, 30)')
2015-11-13 11:22:07,512 12852 INFO o8_exthof_hofermedical werkzeug: 127.0.0.1 - - [13/Nov/2015 11:22:07] "POST /web/dataset/call_button HTTP/1.1" 200 -
```

**Comment by michaelkarrer81** [comment]
@andreparames It might be that i just don't understand how to correctly update a many2one field for a recordset but at least way b seems "valid". Any clarification would be great!

**Comment by andreparames** [comment]
@michaelkarrer81 I should point out I'm not an Odoo employee! Just another partner! But your code seems correct to me, I'll look at the traceback to see if I can understand the problem.

**Comment by andreparames** [comment]
I think the problem is in the module calendar_log_project, in this line: https://github.com/OpenAT/odoo_v8.0/blob/a35295385bedb45e9e84855dee3f9f48f5158c11/addons-own/calendar_log_project/models/calendar.py#L69

You shouldn't use `self.ensure_one()` in the `write()` method, since `self` can have multiple records (like in this case). Instead, you should do a loop that handles each record, like:

    for event in self:
        if event.is_worklog and (event.task_id or event.project_id):
            ...etc...

**Comment by michaelkarrer81** [comment]
@andreparames thank you very much - i close this issue - was clearly my fault and misunderstanding.

**Comment by michaelkarrer81** [comment]
@andreparames I removed my .ensure_one() from the calendar.event write method and made it compatible with record sets  but still 
```
events.write({'category_id': self.category_new_id.id}) 
```
is not working as expected. It will only update one record and leave the others in the recordset unchanged.

You could find the related branch here: 
https://github.com/OpenAT/odoo_v8.0/tree/categories_writefix/addons-own

https://github.com/OpenAT/odoo_v8.0/blob/categories_writefix/addons-own/calendar_log/wizard/wizard.py
https://github.com/OpenAT/odoo_v8.0/blob/categories_writefix/addons-own/calendar_log_project/models/calendar.py

### Example:
```
EVENTS: [655, 654]
Event Name: A 3  Kategory: Kat A
Event Name: A 2  Kategory: Kat A
```
```
events = self.env['calendar.event'].search([('category_id', '=', self.category_old_id.id)])
events.write({'category_id': self.category_new_id.id}) 
```
```
Event Name: A 3  New Kategory: Kat A
Event Name: A 2  New Kategory: ALLGEMEINE TÄTIGKEIT
```

@rco-odoo Is there any reason why recordset.write is not working for all members of the recordset? Is there a basic misunderstandig of .write or @api.multi the cause? What would be the correct way?

**Comment by rco-odoo** [comment]
@michaelkarrer81 the overridden method `write` of `calendar.event` does a rather stupid loop, that converts record ids to integers where it makes sense. Except that the loop is buggy and only keeps the last record id: https://github.com/odoo/odoo/blob/8.0/addons/calendar/calendar.py#L1587

As a temporary workaround, simply update calendar events one at a time.

I will see whether I can fix it.

**Comment by rco-odoo** [comment]
@michaelkarrer81 the issue is fixed in 8.0!
